iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -526,6 +526,68 @@ LogicalResult SortOp::generateScalarImplementation(OpBuilder &b, Location loc,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// FftOp
+//===----------------------------------------------------------------------===//
+
+static LogicalResult verifyFftOp(FftOp op) {
+  auto length = op.getFftLength();
+  if (length == ShapedType::kDynamicSize) return failure();
+  // Expect to be a power of 2.
+  if (length & (length - 1)) return failure();
+  if (!op.getNumInputs() || !op.isScalar(op.getInputOperand(0))) {
+    return op.emitOpError("expected to carry `stage` input");
+  }
+  if (op.getNumOutputs() != 2) {
+    return op.emitOpError("expected outputs to be real and imag tensor/memref");
+  }
+  return success();
+}
+
+void FftOp::getEffects(
```

Comment:
If this is boiler plate, add this as `codeDeclaration` in ODS?

Reply:
I take a look, but still don't know how to make it in ODS. I could not find any examples. The Linalg dialect and IREE VM dialect also implements them in the *.cpp.  Instead, I create a macro for it.

https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/llvm/llvm-project%24+file:%5Emlir/lib/Dialect/Linalg/IR/LinalgOps%5C.cpp+getEffects&patternType=literal

https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/google/iree%24+file:%5Eiree/compiler/Dialect/VM/IR/VMOps%5C.cpp+getEffect&patternType=literal

Reply:
Wont this work https://github.com/google/iree/blob/6d3f2a5768d9197bc1253ed654ab49c3c7f3ec18/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td#L32 ?

Reply:
We will have to get ClassName. I tried `$cppClass` like what we used in verifier/parser/printer, but it could not recognize it.

E.g., I put below snippet to `extraLinalgExtOpClassDeclaration`which did not work.

```cpp
void $cppClass::getEffects( ...
```

Reply:
You need the $cppClass class.  That goes directly into the op declaration. So you should get 

```
class $cppClass {

    void getEffects(...) { ... };
};

```

Reply:
Yes, you're right. We don't need `class $cppClass {` because it will goes into the declaration. I tried to implement getEffects but failed. It will complain about `class member cannot be redeclared`.

This is because the methods are declared by `DeclareOpInterfaceMethods<MemoryEffectsOpInterface`, which inserts `void getEffects(...);` to `LinalgExtOps.h.inc`. If we put the method to `extraLinalgExtOpClassDeclaration` we will get the error.

We should implement it outside the class declaration, ie implementing `void XXXOp::getEffects()` in cpp file.